BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -16,28 +16,120 @@
 #include <Guid/SmmInformationGuid.h>
 #include <Base.h>
 
-typedef UINT32 (*REG_READ)  (UINTN);
-typedef VOID   (*REG_WRITE) (UINTN, UINT32);
-
 #define NUM_TYPE    2
 #define NUM_WIDTH   3
 #define REG_APM_CNT   0xB2
 
-const REG_WRITE  mRegWrite[NUM_TYPE][NUM_WIDTH] = {
-  { (REG_WRITE) MmioWrite8, (REG_WRITE) MmioWrite16, (REG_WRITE) MmioWrite32 },
-  { (REG_WRITE) IoWrite8, (REG_WRITE) IoWrite16, (REG_WRITE) IoWrite32 }
-};
+/* The enum order must align with similiar definition
+   in BlSmmSupport of UEFI Payload */
+typedef enum {
+  REG_TYPE_S3SAVE_MMIO,
+  REG_TYPE_S3SAVE_IO
+} REG_S3SAVE_TYPE;
+
+/**
+  Read register
+
+  Reads a register with register type in Type and length in Width
+
+  @param[in]      Type          type of register
+  @param[in]      Width         width of register
+  @param[in]      Addr          register address
+  @param[in,out]  Val           register value
 
-const REG_READ mRegRead[NUM_TYPE][NUM_WIDTH] = {
-  { (REG_READ) MmioRead8, (REG_READ) MmioRead16, (REG_READ) MmioRead32 },
-  { (REG_READ) IoRead8, (REG_READ) IoRead16, (REG_READ) IoRead32 }
-};
+  @retval         EFI_SUCCESS   read done
+  @retval         others        error
 
-const UINT8 mWidthToIdx[NUM_WIDTH][2] = {
-  { 0, WIDE8  },
-  { 1, WIDE16 },
-  { 2, WIDE32 }
-};
+**/
+static
+EFI_STATUS
+RegRead (
+  IN      UINT8   Type,
+  IN      UINT8   Width,
+  IN      UINTN   Addr,
+  IN OUT  UINT32 *Val
+  )
+{
+  if (Val == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (Type == REG_TYPE_S3SAVE_MMIO) {
+    if (Width == WIDE8) {
+      *Val = MmioRead8 (Addr);
+    } else if (Width == WIDE16) {
+      *Val = MmioRead16 (Addr);
+    } else if (Width == WIDE32) {
+      *Val = MmioRead32 (Addr);
+    } else {
+      return EFI_UNSUPPORTED;
+    }
+  } else if (Type == REG_TYPE_S3SAVE_MMIO) {

```

Comment:
`REG_TYPE_S3SAVE_MMIO -> REG_TYPE_S3SAVE_IO`

Reply:
will fix in next commit. Thanks for pointing out. My bad, big mistake, I didn't check the read result when running a quick verification yesterday.